import sys
from PyQt6 import QtWidgets, uic
from google import genai
from google.genai.errors import APIError
# ğŸŒŸ ìˆ˜ì •: GenerateContentConfigë¥¼ types ëª¨ë“ˆì—ì„œ ëª…ì‹œì ìœ¼ë¡œ ê°€ì ¸ì˜µë‹ˆë‹¤. ğŸŒŸ
from google.genai import types

# =================================================================
# 1. ì„¤ì • ë° ì´ˆê¸°í™”
# =================================================================

# 1. UI íŒŒì¼ ì´ë¦„ ì„¤ì •
UI_FILE_NAME = "Mygemini.ui"

# ğŸŒŸ ì¤‘ìš”: ì‹¤ì œ API í‚¤ë¥¼ ì…ë ¥í–ˆìŠµë‹ˆë‹¤. ì´ í‚¤ë¥¼ ì‹¤ì œ ì‚¬ìš©í•˜ëŠ” í‚¤ë¡œ êµì²´í•˜ì„¸ìš”. ğŸŒŸ
GEMINI_API_KEY = "GEMINI_API_KEY"

# 2. Gemini í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™”
try:
    # API í‚¤ë¥¼ genai.Client()ì— ì§ì ‘ ì „ë‹¬í•©ë‹ˆë‹¤.
    client = genai.Client(api_key=GEMINI_API_KEY)
    MODEL_NAME = "gemini-2.5-flash" # ì‚¬ìš©í•  ëª¨ë¸ ì§€ì •
except ValueError as e:
    print(f"í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì˜¤ë¥˜: API í‚¤ê°€ ì˜ëª»ë˜ì—ˆê±°ë‚˜ ëˆ„ë½ë˜ì—ˆìŠµë‹ˆë‹¤. {e}")
    sys.exit(1)


# =================================================================
# 2. ë©”ì¸ ì• í”Œë¦¬ì¼€ì´ì…˜ í´ë˜ìŠ¤
# =================================================================

class GeminiApp(QtWidgets.QDialog):
    def __init__(self):
        # QDialogë¥¼ ê¸°ë³¸ í´ë˜ìŠ¤ë¡œ ì‚¬ìš©
        super().__init__()
        
        # UI íŒŒì¼ ë¡œë“œ
        try:
            uic.loadUi(UI_FILE_NAME, self)
        except FileNotFoundError:
            print(f"ì˜¤ë¥˜: {UI_FILE_NAME} íŒŒì¼ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤. ê²½ë¡œë¥¼ í™•ì¸í•´ ì£¼ì„¸ìš”.")
            sys.exit(1)
            
        self.setWindowTitle(f"Gemini Q&A - ëª¨ë¸: {MODEL_NAME}")
        
        # 3. ìœ„ì ¯ì„ ëª…ì‹œì ìœ¼ë¡œ ì°¾ìŠµë‹ˆë‹¤ (findChild ì‚¬ìš©ìœ¼ë¡œ UI ë°”ì¸ë”© ì˜¤ë¥˜ ë°©ì§€)
        
        # QPushButton ë° QLineEditëŠ” ìœ í˜•ì„ ì§€ì •í•©ë‹ˆë‹¤.
        self.btnSend = self.findChild(QtWidgets.QPushButton, 'btnSend')
        self.lineEditMyQuestion = self.findChild(QtWidgets.QLineEdit, 'lineEditMyQuestion')
        
        # lblAnswerëŠ” QTextEditì„ ë¨¼ì € ì°¾ê³ , ì‹¤íŒ¨í•˜ë©´ QPlainTextEditì„ ì‹œë„í•©ë‹ˆë‹¤.
        self.lblAnswer = self.findChild(QtWidgets.QTextEdit, 'lblAnswer') 
        if not self.lblAnswer:
            self.lblAnswer = self.findChild(QtWidgets.QPlainTextEdit, 'lblAnswer')
        
        # ìµœì¢…ì ìœ¼ë¡œ ìœ„ì ¯ì´ ëª¨ë‘ ì„±ê³µì ìœ¼ë¡œ ë¡œë“œë˜ì—ˆëŠ”ì§€ í™•ì¸
        if not self.btnSend or not self.lblAnswer or not self.lineEditMyQuestion:
            missing = []
            if not self.btnSend: missing.append("'btnSend' (QPushButton)")
            # lblAnswerì˜ ê²½ìš°, TextEdit ë˜ëŠ” PlainTextEditì¸ì§€ ë‹¤ì‹œ í•œë²ˆ ì‚¬ìš©ìì—ê²Œ ì•ˆë‚´í•©ë‹ˆë‹¤.
            if not self.lblAnswer: missing.append("'lblAnswer' (QTextEdit ë˜ëŠ” QPlainTextEdit ì´ì–´ì•¼ í•©ë‹ˆë‹¤!)")
            if not self.lineEditMyQuestion: missing.append("'lineEditMyQuestion' (QLineEdit)")
            
            print(f"ì¹˜ëª…ì  ì˜¤ë¥˜: UI íŒŒì¼ ({UI_FILE_NAME})ì—ì„œ ë‹¤ìŒ í•„ìˆ˜ ìœ„ì ¯ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤: {', '.join(missing)}")
            sys.exit(1)

        # 4. ìœ„ì ¯ ì´ë²¤íŠ¸ ì—°ê²° ë° ì„¤ì •
        self.btnSend.clicked.connect(self.generate_response)
        
        # QTextEdit/QPlainTextEditì€ setReadOnlyë¥¼ ì§€ì›í•©ë‹ˆë‹¤.
        self.lblAnswer.setReadOnly(True)

        self.show()

    def generate_response(self):
        """
        lineEditMyQuestionì˜ ë‚´ìš©ì„ ê°€ì ¸ì™€ Gemini APIë¥¼ í˜¸ì¶œí•˜ê³ 
        lblAnswerì— ê²°ê³¼ë¥¼ ì§€ì •ëœ í˜•ì‹ìœ¼ë¡œ í‘œì‹œí•©ë‹ˆë‹¤.
        """
        
        # 5. QLineEditì—ì„œ ì‚¬ìš©ì ì§ˆë¬¸ ê°€ì ¸ì˜¤ê¸°
        user_question = self.lineEditMyQuestion.text().strip()
        
        if not user_question:
            self.lblAnswer.setText("ì§ˆë¬¸ì„ ì…ë ¥í•´ ì£¼ì„¸ìš”.")
            return
        
        # ì‘ë‹µ ìƒì„± ì¤‘ì„ì„ ì‚¬ìš©ìì—ê²Œ í‘œì‹œí•˜ê³  UI ì—…ë°ì´íŠ¸
        # ğŸŒŸ ìˆ˜ì •: ì§ˆë¬¸ì„ í‘œì‹œí•˜ëŠ” í…ìŠ¤íŠ¸ì—ì„œ ë³¼ë“œì²´ ë§ˆí¬ë‹¤ìš´(**)ì„ ì œê±°í–ˆìŠµë‹ˆë‹¤.
        self.lblAnswer.setText(f"[ì§ˆë¬¸] {user_question}\n\n[Mygemini] ì‘ë‹µì„ ìƒì„±í•˜ëŠ” ì¤‘ì…ë‹ˆë‹¤...")
        QtWidgets.QApplication.processEvents() 

        # 6. Gemini API í˜¸ì¶œ
        try:
            # ğŸŒŸ ìˆ˜ì •: types.GenerateContentConfigë¥¼ ì‚¬ìš©í•˜ì—¬ ì„¤ì • ê°ì²´ë¥¼ ìƒì„±í•©ë‹ˆë‹¤.
            # ì´ ì„¤ì •ì€ í•œêµ­ì–´ ë‹µë³€ì„ ìœ ë„í•©ë‹ˆë‹¤.
            config = types.GenerateContentConfig(
                system_instruction="You are a helpful assistant. Please answer all questions in Korean."
            )

            response = client.models.generate_content(
                model=MODEL_NAME,
                contents=user_question,
                config=config # ì„¤ì • ì ìš©
            )
            
            # ì‘ë‹µ í…ìŠ¤íŠ¸ ì¶”ì¶œ
            gemini_response = response.text
            
        except APIError as e:
            gemini_response = f"API ì˜¤ë¥˜ ë°œìƒ: Gemini ì„œë²„ì— ì—°ê²°í•  ìˆ˜ ì—†ìŠµë‹ˆë‹¤. (ì˜¤ë¥˜: {e})"
        except Exception as e:
            # API ê´€ë ¨ ì˜¤ë¥˜ ì™¸ì˜ ëª¨ë“  ì˜¤ë¥˜ë¥¼ ì—¬ê¸°ì„œ ì²˜ë¦¬í•©ë‹ˆë‹¤.
            gemini_response = f"ì˜ˆê¸°ì¹˜ ì•Šì€ ì˜¤ë¥˜ ë°œìƒ: {e}"
            
        # 7. ì§€ì •ëœ í˜•ì‹ìœ¼ë¡œ ì¶œë ¥ í…ìŠ¤íŠ¸ í¬ë§·íŒ…
        # ğŸŒŸ ìˆ˜ì •: ì§ˆë¬¸ê³¼ ë‹µë³€ í…ìŠ¤íŠ¸ì—ì„œ ë³¼ë“œì²´ ë§ˆí¬ë‹¤ìš´(**)ì„ ëª¨ë‘ ì œê±°í–ˆìŠµë‹ˆë‹¤.
        formatted_output = (
            f"[ì§ˆë¬¸] {user_question}\n\n"
            f"[Mygemini] {gemini_response}"
        )
        
        # 8. QTextEdit/QPlainTextEditì— í…ìŠ¤íŠ¸ ì„¤ì • (ìŠ¤í¬ë¡¤ ì§€ì›)
        self.lblAnswer.setText(formatted_output)
        
        # 9. QLineEdit ë‚´ìš© ì§€ìš°ê¸°
        self.lineEditMyQuestion.clear()


# =================================================================
# 3. í”„ë¡œê·¸ë¨ ì‹¤í–‰
# =================================================================

if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    window = GeminiApp()

    sys.exit(app.exec())
